### ISSUE TITLE
Better Internet Quest Distribution

### ISSUE THREAD

**Issue by NPBruce** [NPBruce/valkyrie#111]
Quests are downloadable, but want to move to something like ModDB to allow users to log in, rank, comment, browse content.

**Comment by NPBruce** [comment]
From #5

**Comment by NPBruce** [comment]
I've been experimenting with moddb and I can't get it to work well for this function.  Looking to add author, time, difficulty, etc to the quest instead.

It world still be best to have a page to list available quests, post info and get comments. I have submitted Valkyrie as a bgg item as that is one option. Still considering other ideas.

**Comment by redwolf2** [comment]
Sites like minecraftmods.com seem to use wordpress as a CMS.

**Comment by NPBruce** [comment]
I was trying to avoid maintaining such a system.  Now I am working on making the in game system better, maybe just use that for now.

**Comment by facit** [comment]
This feature is so important and has to be really, really easy to use. Both for the creator and for users. The way it is now it's really easy to use for the player, but not easy for creators. Until it is, this program won't get much traction. I can help if you need help but I would need some guidance of where to start. How does the program get the mod files?

**Comment by NPBruce** [comment]
Here is where to start:
https://github.com/NPBruce/valkyrie/wiki/QuestPublish

Valkyrie check the manifest file (https://github.com/NPBruce/valkyrie-store/blob/master/MoM/manifest.ini) and fetches images and packages from there.

Really though Valkyrie can do anything, as long as it can get the info.

The problem is having a web interface that is robust, easy, allows user account but without being annoying, and has all of the modern web features people expect. I still think using an existing service is a good path, but I can't find any that are suitable.  Obviously something custom could be created, but a half job would be worse than what we have currently.

As for the 'not easy for creators' I don't think this is such a big deal.  Creators can just send me a copy and not have to use git.  The big advantage to using a web system is promoting community, allowing discussion, ratings and collaboration.

**Comment by redwolf2** [comment]
For a first step [choose a CMS](https://websitesetup.org/cms-comparison-wordpress-vs-joomla-drupal/). 
Then install something like [WP Download Manager](https://wpdownloadmanager.com) on it (here is a [Sandbox Demo](https://demo.wpdownloadmanager.com/wpdmpro/) (user: demo, password: demo)).
For a forum [Discourse](https://github.com/discourse/discourse) could be used ([Demo here](https://github.com/discourse/discourse)) it also comes with a wordpress integration.

I'd say worry about the valkyrie store integration later. 

It could just be a second repository url to a [httpd directory listing](https://stackoverflow.com/questions/39239276/how-to-enable-directory-listing-in-apache-web-server) service.
The result is something like this:
http://srp.nu/

**Comment by ChadMcCallum** [comment]
I'm actually a web developer by trade and would be able to contribute to this, although it really should be its own repo with features and whatnot (linking back to the manifest file in this repo).  Is there a centralized document / communication platform you guys use for planning?

**Comment by NPBruce** [comment]
I completely agree it should be it's own repo if that is the solution.  There is no planned communication, but implementing on the Valkyrie side really isn't hard, just need to be able to do a public http request for:

* Quest list
* Meta data for each quest
* Package file

It may get a little trickier in the future for content pack support which could be a lot bigger, but it wouldn't change much.

I would still like to use an off the shelf solution as much as possible, I don't want to maintain another project.

**Comment by massimoregoli** [comment]
Any news about the web interface guys? Do you still need help to find a good solution? I think it would be great to have it, not just for scenarios rating, but because it would be a good way to attract people.

**Comment by NPBruce** [comment]
No good solution has been found at this point.

**Comment by eorahil** [comment]
My scenario is near to an end and I decided to create a google docs poll in order to let people rate it in a centralized way. The link to the poll will be available in the scenario itself.

**Comment by scrubbless** [comment]
A central place for rating and feed back would be nice, though its important that it is not half hearted. As it will be a platform to build a community.

**Comment by BenRQ** [comment]
Hi all,
I see this ticket as been voted as quite important, and it's true it would help to build a community to have better communication between players and authors.

But technical implementation is not moving much, mainly because of the amount of different request stacked on this topic. 
I agree it would be great to do it all, but I would suggest to do it step by step.

I would divide all requests into these tasks/topics:
 1) be able to rank / provide simple feedback for a scenario that has been played (I would suggest from the app once the scenario is finished)
 2) Present the scenarios with the rank / number of voters
 3) Present a better description / additional metadata and allow filter (rank, author, languages, number of time played, average duration ...) #832
 4) Have an easy(easier?) way to push new scenario.
 5) have a portal to enable discussion between authors and players ( discord? reddit? ) #860
 6) Have a website with all of the metadata available for each scenarios
 
1 and 2 would be top priority I think.
5 would probably be the easiest thing to do.

@redwolf2, I think your proposal for implementation is very good.
I would be ready to take part in the coding, but I have one main question: where would we host this ?
It would have to be free, reliable (and last for a few years).

Otherwise, I was thinking about google forms / google spreadsheet.
It would not be as efficient, as flexible as a database, but it would be quite easy to implement, and reliable.
How do you feel about using google platform for this?

@NPBruce, I will not work on this before I have your go on this :)
And maybe you prefer working on this yourself ?

**Comment by NPBruce** [comment]
I don't have capacity to work on this (as you have no doubt noticed I'm not on top of core development).

I am happy for anyone else to take this on. I don't entirely agree that doing things inside the app is the first step. I would suggest starting a web interface first, even if very basic, and the form that it takes will shape how the data is stored and how the app interacts.

Though I guess you can just move it later. Either way, the issue with feedback in the app is how do you authenticate the write to whereever the data is stored? Perhaps something as simple as sending an email to a specific account with a volunteer who can manually vet the reponses will do for now? The email could still be checked by a script later.

**Comment by BenRQ** [comment]
Hi @NPBruce ,
I'm happy to help and I have to say that the code is really well thought and structured.
It's really nice to work on this project also for this.

To answer your question:
> I would suggest starting a web interface first, even if very basic, and the form that it takes will shape how the data is stored and how the app interacts.

I agree, and I have done a first draft of Google forms.
Here it is : https://goo.gl/forms/sCiwEojsoK6yWNSk1
feel free to have a look and comment
This form will be almost fully filled automatically by the app, except for the field marked as 'manual'.

>Either way, the issue with feedback in the app is how do you authenticate the write to wherever the data is stored?

The form would have to be without authentication.
But as it would be called from the app, the URL will not be shared.
It would still be in the code, but we can add one or two points of security, just to discourage the few ill intended geek.

> The email could still be checked by a script later.

All scripted actions would require a specific hosting.
If we have a specific hosting, I would rather use a db than a google forms + google sheets.
But I don't see an easy way to get that.

**Comment by NPBruce** [comment]
Assuming you can submit google form data with a simple http request that look easy enough.

**Comment by BenRQ** [comment]
Yes, it should be easy enough with a single HTTP Post request.
Form result will be pushed to a spreadsheet that will have its own script attached.
This script will generate a JSON with the rating, numbers of play session for each scenario. Generation will be done automatically.
Json file will be free to access, but read-only.

Full spreadsheet will be available in read-only to authors, via authentication.

It seems feasible, now I will see how easy it is when I will start working on it :)

One thing I don't know is how to detect that the scenario is ended.
Is there a way to do this, so I can display rating popup at the right time ?
Or should I do it every time players leave the session?

**Comment by BenRQ** [comment]
I've been spending some time on this : 
 - unofficial Valkyrie google account created
 - Google form : DONE
 - Google sheet shareable file with live results : AVAILABLE (can be shared with authors in read-only)
 - Google script to generate JSON, every hour : DONE (JSON is available on a google drive)
 - Class creation to push content from Valkyrie to Google form : DONE (with Unity lib)
 - New 'rating' screen displayed at end of scenario : ONGOING (screen is displayed at the the right time, thanks Bruce for the tip, design implementation ongoing with 3 field : did you won ? your rating? Comments ?)
 - Gathering data from ongoing QUEST to push to the form automatically : Almost DONE (only game duration is missing)

Not started yet:
 - download JSON from Google drive (should be easy)
 - Display the data coming from the JSON in the scenario list (download, and quest selection) 
**if anyone has some suggestions on how to display the following information on those two screens:**
rating, number of time played, average duration, average win ratio

I will be out of home for two weeks, and I plan to work on this when I'm back.

Any design suggestion would help me a lot :)

**Comment by NPBruce** [comment]
Asking if you won is interesting, as 'winning' is not a concept Valkyrie is aware of, there may be different endings.

I think including the log would be a great help for scenario developers, they could see exactly what choices are made and what happened. This may be what you mean by 'data from ongoing QUEST'.

**Comment by BenRQ** [comment]
Yes exactly, all player choices will be recorded and sent to the form, and it will be available to quest authors.

Also the number of players, name of investigators, language and of course scenario id.

Scenario id will be the filename as I think it is unique and it does not change with the language selected.

**Comment by BenRQ** [comment]
Here is a preview of the rating screen.
Please tell me if you have any suggestion for improvements : 

![image](https://user-images.githubusercontent.com/40794717/44303188-45d7c780-a33a-11e8-8f64-6bf2bf91dcb5.png)

**Comment by BenRQ** [comment]
Hello
I just finished implementing the core functions of the rating system in Valkyrie
Design is not final yet, it is just here for test purposes (I don't really like the 'hearts')
but here is the quest list with rating, , number of play session, average win ratio: 
https://cdn.discordapp.com/attachments/473586525477666826/480880321806598175/unknown.png

And the end game rating form look like this :
https://cdn.discordapp.com/attachments/473586525477666826/480880685348028446/unknown.png

Any help/suggestion for the design of the scenario selection screen with rating, number of play session, average win ratio would be welcome
For the end game form, i think it is OK as it is for a V1, but any suggestion for improvement would be welcome too.

Feel free to join on discord, if you want to discuss this directly :)

**Comment by BenRQ** [comment]
The ratings are now available :  
![image](https://user-images.githubusercontent.com/40794717/44633542-fffbae80-a98c-11e8-93d9-beab66f9d4ac.png)

Even with some tentacles !

Please join in the Discord if you would like to alpha test the version.

**Comment by BenRQ** [comment]
Beta test version available with implementation of this issue:
https://github.com/BenRQ/valkyrie/releases/tag/release%2F2.1%2FBrynhildr_2.1.0b

You can see rating when browsing quests.
You can rate, comment a scenario at the end of the game.
Authors can access statistics and comments from player in a Google sheet.

To access the file, you should create a new issue and tag me, I will share the read-only access to the file.

@NPBruce @redwolf2 tell me if I can close this issue or if you want to keep it open until all points have been addressed.
